File: facebook-page-post/scripts/post.py
#!/usr/bin/env python3
"""
Facebook Page Post - Python 腳本
使用 Facebook Graph API v24.0 發布文字、圖片、影片到 Facebook 粉絲專頁
"""

import os
import sys
import requests
import argparse
import time
from pathlib import Path

# Configuration
PAGE_ID = os.environ.get("FACEBOOK_PAGE_ID")
ACCESS_TOKEN = os.environ.get("FACEBOOK_PAGE_ACCESS_TOKEN")
API_VERSION = "v24.0"
BASE_URL = f"https://graph.facebook.com/{API_VERSION}"

# Retry settings
MAX_RETRIES = 3
RETRY_DELAY = 5  # seconds


def validate_credentials():
    """驗證環境變數"""
    if not PAGE_ID:
        print("錯誤: 請設定 FACEBOOK_PAGE_ID 環境變數", file=sys.stderr)
        print("範例: export FACEBOOK_PAGE_ID=\"123456789\"", file=sys.stderr)
        print("詳細說明請參考: https://developers.facebook.com/docs/pages/access-tokens/", file=sys.stderr)
        sys.exit(1)

    if not ACCESS_TOKEN:
        print("錯誤: 請設定 FACEBOOK_PAGE_ACCESS_TOKEN 環境變數", file=sys.stderr)
        print("範例: export FACEBOOK_PAGE_ACCESS_TOKEN=\"EAAxxxxxx...\"", file=sys.stderr)
        print("取得方式請參考 SKILL.md 的 Token 管理章節", file=sys.stderr)
        sys.exit(1)


def make_api_request(url, data=None, files=None, max_retries=MAX_RETRIES):
    """
    發送 API 請求，包含重試機制

    Args:
        url: API URL
        data: POST data dict
        files: files dict for multipart upload
        max_retries: 最大重試次數

    Returns:
        API 回應 JSON
    """
    for attempt in range(max_retries):
        try:
            if files:
                response = requests.post(url, data=data, files=files)
            else:
                response = requests.post(url, json=data)

            # 檢查是否為 Rate Limit (368)
            if response.status_code == 200 or response.status_code == 201:
                return response.json()

            # 檢查是否為 Token 過期
            if response.status_code == 401:
                error_data = response.json()
                if error_data.get('error', {}).get('code') == 190:
                    print("⚠️  錯誤: Access Token 已過期或無效", file=sys.stderr)
                    print("請使用以下指令檢查 Token 狀態:", file=sys.stderr)
                    print("  python3 .claude/skills/facebook-page-post/scripts/token-helper.py", file=sys.stderr)
                    print("並參考 SKILL.md 更新 Token", file=sys.stderr)
                    sys.exit(1)

            # 檢查是否為 Rate Limit
            if response.status_code == 429 or \
               (response.status_code == 400 and response.json().get('error', {}).get('code') == 368):
                if attempt < max_retries - 1:
                    print(f"⏳ 達到速率限制，等待 {RETRY_DELAY} 秒後重試... (嘗試 {attempt + 1}/{max_retries})")
                    time.sleep(RETRY_DELAY)
                    continue

            # 其他錯誤
            response.raise_for_status()
            return response.json()

        except requests.exceptions.RequestException as e:
            if attempt < max_retries - 1:
                print(f"⚠️  請求失敗: {e}，重試中... (嘗試 {attempt + 1}/{max_retries})")
                time.sleep(RETRY_DELAY)
                continue
            else:
                print(f"❌ API 請求失敗: {e}", file=sys.stderr)
                sys.exit(1)


def post_text(message):
    """
    發布文字文章到 Facebook Page

    Args:
        message: 文字內容

    Returns:
        API 回應，包含貼文 ID
    """
    if not message:
        print("錯誤: 文字文章需要 --message 參數", file=sys.stderr)
        sys.exit(1)

    url = f"{BASE_URL}/{PAGE_ID}/feed"
    payload = {
        "message": message,
        "access_token": ACCESS_TOKEN
    }

    result = make_api_request(url, data=payload)

    print(f"✅ 文章發布成功!")
    print(f"   貼文 ID: {result.get('id')}")
    print(f"   連結: https://www.facebook.com/{result.get('id').split('_')[1]}")

    return result


def post_photo(image_path, message=None):
    """
    上傳單張圖片到 Facebook Page

    Args:
        image_path: 圖片檔案路徑
        message: 圖片說明（可選）

    Returns:
        API 回應，包含圖片 ID
    """
    if not image_path:
        print("錯誤: 圖片文章需要 --file 參數", file=sys.stderr)
        sys.exit(1)

    if not os.path.exists(image_path):
        print(f"錯誤: 檔案不存在 - {image_path}", file=sys.stderr)
        sys.exit(1)

    url = f"{BASE_URL}/{PAGE_ID}/photos"
    payload = {"access_token": ACCESS_TOKEN}

    if message:
        payload["caption"] = message

    with open(image_path, "rb") as f:
        files = {"source": f}
        result = make_api_request(url, data=payload, files=files)

    print(f"✅ 圖片上傳成功!")
    print(f"   圖片 ID: {result.get('id')}")
    if result.get('post_id'):
        print(f"   貼文連結: https://www.facebook.com/{result.get('post_id').split('_')[1]}")

    return result


def post_photos(image_paths, message=None):
    """
    上傳多張圖片到 Facebook Page（會建立相簿）

    Args:
        image_paths: 圖片檔案路徑列表
        message: 相簿標題（可選）

    Returns:
        API 回應列表
    """
    if not image_paths or len(image_paths) == 0:
        print("錯誤: 多圖上傳需要 --files 參數", file=sys.stderr)
        sys.exit(1)

    # 檢查所有檔案是否存在
    for img_path in image_paths:
        if not os.path.exists(img_path):
            print(f"錯誤: 檔案不存在 - {img_path}", file=sys.stderr)
            sys.exit(1)

    print(f"📁 正在建立相簿並上傳 {len(image_paths)} 張圖片...")

    # 建立未發布的相簿
    album_url = f"{BASE_URL}/{PAGE_ID}/albums"
    album_payload = {
        "name": message or "Photo Album",
        "access_token": ACCESS_TOKEN
    }

    album_result = make_api_request(album_url, data=album_payload)
    album_id = album_result.get("id")

    print(f"   相簿 ID: {album_id}")

    # 上傳圖片到相簿
    results = []
    for i, image_path in enumerate(image_paths, 1):
        photo_url = f"{BASE_URL}/{album_id}/photos"

        with open(image_path, "rb") as f:
            files = {"source": f}
            result = make_api_request(photo_url, data={"access_token": ACCESS_TOKEN}, files=files)
            results.append(result)

        print(f"   圖片 {i}/{len(image_paths)} 上傳完成 (ID: {result.get('id')})")

    print(f"✅ 所有圖片上傳成功!")
    print(f"   相簿連結: https://www.facebook.com/media/set/?set={album_id}")

    return results


def post_video(video_path, message=None, title=None, description=None):
    """
    上傳影片到 Facebook Page

    Args:
        video_path: 影片檔案路徑
        message: 影片描述（可選）
        title: 影片標題（可選）
        description: 影片詳細描述（可選）

    Returns:
        API 回應，包含影片 ID
    """
    if not video_path:
        print("錯誤: 影片上傳需要 --file 參數", file=sys.stderr)
        sys.exit(1)

    if not os.path.exists(video_path):
        print(f"錯誤: 檔案不存在 - {video_path}", file=sys.stderr)
        sys.exit(1)

    # 檢查檔案大小
    file_size = os.path.getsize(video_path)
    file_size_mb = file_size / (1024 * 1024)

    if file_size_mb > 1000:
        print(f"⚠️  警告: 影片大小 {file_size_mb:.1f} MB，超過 1GB", file=sys.stderr)
        print("建議使用 Facebook 的 Resumable Upload API 上傳大型影片", file=sys.stderr)
        response = input("是否繼續? (y/n): ")
        if response.lower() != 'y':
            sys.exit(1)

    url = f"{BASE_URL}/{PAGE_ID}/videos"
    payload = {"access_token": ACCESS_TOKEN}

    # 設定影片資訊
    if message:
        payload["description"] = message
    if title:
        payload["title"] = title
    if description:
        payload["description"] = description

    print(f"📹 正在上傳影片... (大小: {file_size_mb:.1f} MB)")

    with open(video_path, "rb") as f:
        files = {"source": f}
        result = make_api_request(url, data=payload, files=files)

    print(f"✅ 影片上傳成功!")
    print(f"   影片 ID: {result.get('id')}")

    # 影片需要處理時間
    print(f"   注意: 影片正在處理中，請稍候片刻後查看")

    return result


def main():
    """主程式"""
    parser = argparse.ArgumentParser(
        description="Facebook Page Post - 發布文章、圖片、影片到 Facebook 粉絲專頁",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
範例:
  文字文章:
    python3 %(prog)s text --message "Hello, World!"

  單張圖片:
    python3 %(prog)s photo --file photo.jpg --message "Beautiful photo"

  多張圖片:
    python3 %(prog)s photos --files photo1.jpg photo2.jpg --message "Album"

  影片上傳:
    python3 %(prog)s video --file video.mp4 --title "My Video"

環境變數:
  FACEBOOK_PAGE_ID         Facebook 粉絲專頁 ID
  FACEBOOK_PAGE_ACCESS_TOKEN   Facebook Page Access Token

詳細說明請參考 SKILL.md
        """
    )

    parser.add_argument(
        "type",
        choices=["text", "photo", "photos", "video"],
        help="發文類型"
    )
    parser.add_argument(
        "--message",
        help="文字內容或圖片/影片說明"
    )
    parser.add_argument(
        "--file",
        help="單一檔案路徑（用於 photo 或 video）"
    )
    parser.add_argument(
        "--files",
        nargs="+",
        help="多個檔案路徑（用於 photos）"
    )
    parser.add_argument(
        "--title",
        help="影片標題"
    )
    parser.add_argument(
        "--description",
        help="影片詳細描述"
    )

    args = parser.parse_args()

    # 驗證環境變數
    validate_credentials()

    try:
        # 根據類型執行對應功能
        if args.type == "text":
            post_text(args.message)

        elif args.type == "photo":
            post_photo(args.file, args.message)

        elif args.type == "photos":
            post_photos(args.files, args.message)

        elif args.type == "video":
            # 合併 message 和 description
            description = args.description or args.message
            post_video(args.file, description, args.title, description)

    except KeyboardInterrupt:
        print("\n\n⚠️  操作已取消")
        sys.exit(1)
    except Exception as e:
        print(f"❌ 發生錯誤: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: facebook-page-post/scripts/post.js
#!/usr/bin/env node
/**
 * Facebook Page Post - Node.js 腳本
 * 使用 Facebook Graph API v24.0 發布文字、圖片、影片到 Facebook 粉絲專頁
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const querystring = require('querystring');

// Configuration
const PAGE_ID = process.env.FACEBOOK_PAGE_ID;
const ACCESS_TOKEN = process.env.FACEBOOK_PAGE_ACCESS_TOKEN;
const API_VERSION = 'v24.0';
const BASE_URL = `https://graph.facebook.com/${API_VERSION}`;

// Retry settings
const MAX_RETRIES = 3;
const RETRY_DELAY = 5000; // milliseconds

/**
 * Validate environment variables
 */
function validateCredentials() {
    if (!PAGE_ID) {
        console.error('錯誤: 請設定 FACEBOOK_PAGE_ID 環境變數');
        console.error('範例: export FACEBOOK_PAGE_ID="123456789"');
        console.error('詳細說明請參考: https://developers.facebook.com/docs/pages/access-tokens/');
        process.exit(1);
    }

    if (!ACCESS_TOKEN) {
        console.error('錯誤: 請設定 FACEBOOK_PAGE_ACCESS_TOKEN 環境變數');
        console.error('範例: export FACEBOOK_PAGE_ACCESS_TOKEN="EAAxxxxxx..."');
        console.error('取得方式請參考 SKILL.md 的 Token 管理章節');
        process.exit(1);
    }
}

/**
 * Sleep utility
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Make HTTP POST request
 */
function makeRequest(url, data, files = null) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);

        if (files) {
            // Multipart form data upload
            const boundary = '----WebKitFormBoundary' + Date.now();
            const body = [];

            // Add form fields
            for (const [key, value] of Object.entries(data)) {
                body.push(`--${boundary}\r\n`);
                body.push(`Content-Disposition: form-data; name="${key}"\r\n\r\n`);
                body.push(`${value}\r\n`);
            }

            // Add file
            if (files.source) {
                body.push(`--${boundary}\r\n`);
                body.push(`Content-Disposition: form-data; name="source"; filename="${files.filename}"\r\n`);
                body.push(`Content-Type: application/octet-stream\r\n\r\n`);
                body.push(files.source);
                body.push('\r\n');
            }

            body.push(`--${boundary}--\r\n`);

            const options = {
                hostname: urlObj.hostname,
                port: 443,
                path: urlObj.pathname + urlObj.search,
                method: 'POST',
                headers: {
                    'Content-Type': `multipart/form-data; boundary=${boundary}`,
                    'Content-Length': Buffer.concat(body.map(b =>
                        Buffer.isBuffer(b) ? b : Buffer.from(b)
                    )).length
                }
            };

            const req = https.request(options, (res) => {
                let responseData = '';
                res.on('data', (chunk) => responseData += chunk);
                res.on('end', () => {
                    try {
                        const jsonResponse = JSON.parse(responseData);
                        if (res.statusCode === 200 || res.statusCode === 201) {
                            resolve(jsonResponse);
                        } else {
                            reject(new Error(`HTTP ${res.statusCode}: ${JSON.stringify(jsonResponse)}`));
                        }
                    } catch (e) {
                        reject(new Error(`Failed to parse response: ${responseData}`));
                    }
                });
            });

            req.on('error', reject);

            // Send body
            const buffers = body.map(b => Buffer.isBuffer(b) ? b : Buffer.from(b));
            req.write(Buffer.concat(buffers));
            req.end();
        } else {
            // JSON request
            const postData = JSON.stringify(data);

            const options = {
                hostname: urlObj.hostname,
                port: 443,
                path: urlObj.pathname + urlObj.search,
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Length': Buffer.byteLength(postData)
                }
            };

            const req = https.request(options, (res) => {
                let responseData = '';
                res.on('data', (chunk) => responseData += chunk);
                res.on('end', () => {
                    try {
                        const jsonResponse = JSON.parse(responseData);
                        if (res.statusCode === 200 || res.statusCode === 201) {
                            resolve(jsonResponse);
                        } else {
                            reject(new Error(`HTTP ${res.statusCode}: ${JSON.stringify(jsonResponse)}`));
                        }
                    } catch (e) {
                        reject(new Error(`Failed to parse response: ${responseData}`));
                    }
                });
            });

            req.on('error', reject);
            req.write(postData);
            req.end();
        }
    });
}

/**
 * Make API request with retry logic
 */
async function makeApiRequest(url, data, files = null) {
    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
        try {
            const result = await makeRequest(url, data, files);
            return result;
        } catch (error) {
            if (attempt < MAX_RETRIES) {
                console.log(`⚠️  請求失敗: ${error.message}，重試中... (嘗試 ${attempt}/${MAX_RETRIES})`);
                await sleep(RETRY_DELAY);
            } else {
                console.error(`❌ API 請求失敗: ${error.message}`);
                process.exit(1);
            }
        }
    }
}

/**
 * Post text message
 */
async function postText(message) {
    if (!message) {
        console.error('錯誤: 文字文章需要 --message 參數');
        process.exit(1);
    }

    const url = `${BASE_URL}/${PAGE_ID}/feed`;
    const payload = {
        message: message,
        access_token: ACCESS_TOKEN
    };

    const result = await makeApiRequest(url, payload);

    console.log('✅ 文章發布成功!');
    console.log(`   貼文 ID: ${result.id}`);
    console.log(`   連結: https://www.facebook.com/${result.id.split('_')[1]}`);

    return result;
}

/**
 * Post photo
 */
async function postPhoto(imagePath, message = null) {
    if (!imagePath) {
        console.error('錯誤: 圖片文章需要 --file 參數');
        process.exit(1);
    }

    if (!fs.existsSync(imagePath)) {
        console.error(`錯誤: 檔案不存在 - ${imagePath}`);
        process.exit(1);
    }

    const url = `${BASE_URL}/${PAGE_ID}/photos`;
    const payload = {
        access_token: ACCESS_TOKEN
    };

    if (message) {
        payload.caption = message;
    }

    const imageBuffer = fs.readFileSync(imagePath);

    const files = {
        source: imageBuffer,
        filename: path.basename(imagePath)
    };

    const result = await makeApiRequest(url, payload, files);

    console.log('✅ 圖片上傳成功!');
    console.log(`   圖片 ID: ${result.id}`);
    if (result.post_id) {
        console.log(`   貼文連結: https://www.facebook.com/${result.post_id.split('_')[1]}`);
    }

    return result;
}

/**
 * Post multiple photos (album)
 */
async function postPhotos(imagePaths, message = null) {
    if (!imagePaths || imagePaths.length === 0) {
        console.error('錯誤: 多圖上傳需要 --files 參數');
        process.exit(1);
    }

    // Check all files exist
    for (const imgPath of imagePaths) {
        if (!fs.existsSync(imgPath)) {
            console.error(`錯誤: 檔案不存在 - ${imgPath}`);
            process.exit(1);
        }
    }

    console.log(`📁 正在建立相簿並上傳 ${imagePaths.length} 張圖片...`);

    // Create album
    const albumUrl = `${BASE_URL}/${PAGE_ID}/albums`;
    const albumPayload = {
        name: message || 'Photo Album',
        access_token: ACCESS_TOKEN
    };

    const albumResult = await makeApiRequest(albumUrl, albumPayload);
    const albumId = albumResult.id;

    console.log(`   相簿 ID: ${albumId}`);

    // Upload photos
    const results = [];
    for (let i = 0; i < imagePaths.length; i++) {
        const imagePath = imagePaths[i];
        const photoUrl = `${BASE_URL}/${albumId}/photos`;

        const imageBuffer = fs.readFileSync(imagePath);
        const files = {
            source: imageBuffer,
            filename: path.basename(imagePath)
        };

        const result = await makeApiRequest(photoUrl, { access_token: ACCESS_TOKEN }, files);
        results.push(result);

        console.log(`   圖片 ${i + 1}/${imagePaths.length} 上傳完成 (ID: ${result.id})`);
    }

    console.log('✅ 所有圖片上傳成功!');
    console.log(`   相簿連結: https://www.facebook.com/media/set/?set=${albumId}`);

    return results;
}

/**
 * Post video
 */
async function postVideo(videoPath, message = null, title = null, description = null) {
    if (!videoPath) {
        console.error('錯誤: 影片上傳需要 --file 參數');
        process.exit(1);
    }

    if (!fs.existsSync(videoPath)) {
        console.error(`錯誤: 檔案不存在 - ${videoPath}`);
        process.exit(1);
    }

    // Check file size
    const stats = fs.statSync(videoPath);
    const fileSizeMB = stats.size / (1024 * 1024);

    if (fileSizeMB > 1000) {
        console.warn(`⚠️  警告: 影片大小 ${fileSizeMB.toFixed(1)} MB，超過 1GB`);
        console.warn('建議使用 Facebook 的 Resumable Upload API 上傳大型影片');
        // Continue anyway
    }

    const url = `${BASE_URL}/${PAGE_ID}/videos`;
    const payload = {
        access_token: ACCESS_TOKEN
    };

    if (message) {
        payload.description = message;
    }
    if (title) {
        payload.title = title;
    }
    if (description) {
        payload.description = description;
    }

    console.log(`📹 正在上傳影片... (大小: ${fileSizeMB.toFixed(1)} MB)`);

    const videoBuffer = fs.readFileSync(videoPath);
    const files = {
        source: videoBuffer,
        filename: path.basename(videoPath)
    };

    const result = await makeApiRequest(url, payload, files);

    console.log('✅ 影片上傳成功!');
    console.log(`   影片 ID: ${result.id}`);
    console.log(`   注意: 影片正在處理中，請稍候片刻後查看`);

    return result;
}

/**
 * Main function
 */
async function main() {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        console.log('Facebook Page Post - 發布文章、圖片、影片到 Facebook 粉絲專頁\n');
        console.log('使用方法:');
        console.log('  node post.js text --message "Hello"');
        console.log('  node post.js photo --file photo.jpg --message "Caption"');
        console.log('  node post.js photos --files p1.jpg p2.jpg --message "Album"');
        console.log('  node post.js video --file video.mp4 --title "Title"\n');
        console.log('環境變數:');
        console.log('  FACEBOOK_PAGE_ID              Facebook 粉絲專頁 ID');
        console.log('  FACEBOOK_PAGE_ACCESS_TOKEN    Facebook Page Access Token\n');
        console.log('詳細說明請參考 SKILL.md');
        process.exit(1);
    }

    validateCredentials();

    const type = args[0];
    const message = args.includes('--message') ? args[args.indexOf('--message') + 1] : null;
    const file = args.includes('--file') ? args[args.indexOf('--file') + 1] : null;
    const filesIndex = args.indexOf('--files');
    const files = filesIndex !== -1 ? args.slice(filesIndex + 1) : null;
    const title = args.includes('--title') ? args[args.indexOf('--title') + 1] : null;
    const description = args.includes('--description') ? args[args.indexOf('--description') + 1] : null;

    try {
        switch (type) {
            case 'text':
                await postText(message);
                break;
            case 'photo':
                await postPhoto(file, message);
                break;
            case 'photos':
                await postPhotos(files, message);
                break;
            case 'video':
                await postVideo(file, description || message, title, description);
                break;
            default:
                console.error(`錯誤: 未知的類型 "${type}"`);
                console.error('支援的類型: text, photo, photos, video');
                process.exit(1);
        }
    } catch (error) {
        console.error(`❌ 發生錯誤: ${error.message}`);
        process.exit(1);
    }
}

// Run
main().catch(console.error);
